New PR

LOG91 opened the pull request:
다시 PR합니다~

Commit:
Modify publishNumber method and delete lottoArray in buyLottos method

Commit:
Modify switch to object in setLuckyNumber method

Commit:
Modify check lotto method

내가 산 로또 배열들을 원소대로 setLuckyNumber에 들어가는 내가 정한 6개의 숫자와 indexOf를 이용해
일일이 돌아줬던 방식을 setLuckyNumber에 처음 들어올 때에 객체로 바꿔주고 내 로또 배열의 원소별로
객체의 키값에 넣어주는 for문을 사용하여 키값이 있으면 카운트해서 넘겨주는 방식으로 바꿨다

crongro commented on lotto.js:
```diff
@@ -24,64 +24,59 @@ function buyLottos(money) {
-  myTotalPrize += foundThree * PRIZE_MONEY.THREE + foundFour * PRIZE_MONEY.FOUR +
-    foundFive * PRIZE_MONEY.FIVE + foundSix * PRIZE_MONEY.SIX;
-  printResult(foundThree, foundFour, foundFive, foundSix);
+  myTotalPrize += countArr[3] * PRIZE_MONEY.THREE + countArr[4] * PRIZE_MONEY.FOUR +
```
이 코드를 반복을 돌면서 처리할 수 있지 않을까요?

crongro commented on lotto.js:
```diff
@@ -24,64 +24,59 @@ function buyLottos(money) {
 
+// 로또 체크 함수
+checkNumber = (comNum, myNum) => {
+  return myNum.reduce((prev, curr) => {
```
reduce를 잘 썼군요.

crongro commented on lotto.js:
```diff
@@ -24,64 +24,59 @@ function buyLottos(money) {
 }
 
 // 임의로 여섯 개 숫자의 배열을 넣고 실행하면 배열값과 내가 산 로또들과 비교해서 당첨을 결정한다
 let setLuckyNumber = array => {
```
이 함수안에서 동작하는 코드를 함수로 좀더 분리해보면 좋겠어요.
그 크기를 임의로 고민해서 분리해보세요.

crongro commented on lotto.js:
```diff
@@ -24,64 +24,59 @@ function buyLottos(money) {
+
+}
 // 수익률을 계산
 function calculateEarningRate(totalPrizeMoney, investMoney) {
```
화살표함수와 함수선언문을 혼용해서 사용하는 건 학습적인 이유가 아니면 좀 자제해야 할 거 같아요.
일관된 방법으로 구현하세요.

LOG91 replied:
네 ES6로 변경하겠습니다!

crongro reviewed (changes requested):
리뷰남깁니다.
개선을 좀 해보죠.
그외에도 본인이 개선해야 싶은게 보이면 개선해보세요.

Commit:
Add learning keyword markdown

Commit:
Modify myTotalPrize

나의 총 당첨금액을 저장하는 myTotalPrize를 for문을 이용하여 저장

Commit:
Modify method get myTotalPrize

for문을 이용하는 방법에서 reduce를 이용하는 것으로 변경해보았습니다.

Commit:
Modify setLuckyNumber and change syntax ES5 to ES6

1. setLuckyNumber 함수 안에 인자로 들어오는 array길이에 맞춰 가변적으로
맞춘 번호 저장하는 객체 생성

2. 함수 선언을 ES6로 통일

crongro reviewed (changes requested):
내용정리는 잘했고요.

어떤 상황에서 쓰면 유용할 것 같다(또는 주의해야할 거 같다등)는 내용을 항목별 맨 아래에 추가로 써보실래요?  
느낀점을 통해서 좀더 이해도를 높이면 좋겠어요.

Commit:
Add number1 comment

Commit:
Add Number2 comment

Commit:
Add comment number3 to number5

Commit:
Add comment number 6

Commit:
Add comment number 7

Commit:
Add comment number 8

Commit:
Add comment number 9

Commit:
Add comment number 10

Commit:
Add comment number 11

Commit:
Add comment number 12

Commit:
Add comment number 13

Commit:
Add comment number 15

crongro commented on keyword.md:
```diff
@@ -21,6 +21,11 @@
+비트연산자를 공부하며 적을 때까지만 해도 이게 뭔지는 알 것 같은데 어떻게 사용하는지를 몰랐었는데  
+알고리즘 문제를 풀다가 reduce와 XOR를 이용하면 단일한 요소를 구할 수 있는 함수를 구현할 수 있는 것을 보며  
+비트연산자가 괜히 있는 것이 아니구나라는 것을 깨달았다  
+아직 XOR의 예밖에 모르지만 앞으로 나온다면 적극적으로 배우고 싶다!
```
비트연산자로만 할 수 있는 구현이 있다기보다, bit연산자로도 구현할 수 있는 것들이 있긴해요.
간단한 연산이나 비교등에서 유리할 듯.

이런거~
var isOdd = number & 1;

crongro commented on keyword.md:
```diff
@@ -52,6 +62,10 @@ var showName2 = function(name) {
+__Comment__  
+이것을 배우며 이것을 이용하면 넘어오는 값의 boolean값을 판별할 수 있지 않을까  
+하며 if에 사용하면 좋겠다 싶었는데 생각해보니 if는 이미 이 기능을 하고 있는 것 같다  
+if내부에 !!같은게 있지 않을까 생각이 든다
```
네 그럴 수 있어요~ true인지 false인지를 판단하느거니까 그런생각을 가질 수 있겠네요.

crongro commented on keyword.md:
```diff
@@ -79,6 +96,9 @@ switch(name) {
 
+__Comment__  
+JS에서는 === 사용이 거의 대부분일 것 같다 일단 타입을 빼고 값만 들어오면 되는 == 가 필요한 경우도 있을 것 같긴한데  
+무언가 아직은 안전하지 않은 방법이 않을까 생각이 든다
```
==의 특성을 이용하는 경우가  아니면 === 를 쓰는게 정확하겠죠?

crongro commented on keyword.md:
````diff
@@ -92,7 +112,8 @@ function test(e) {
 ```
-
+__Comment__  
+||을 습관화해서 사용한다면 훨씬 더 코드를 깔끔하게 만들 수 있고 가독성에도 훨씬 좋을 것 같다  
````
참고로 es6의 default parameter라는 것이 이부분을 대체할 수 있습니다.

crongro commented on keyword.md:
```diff
@@ -129,6 +150,10 @@ document.write(myBirth);
+__Comment__  
+솔직하게 이 함수는 이미 내 마음을 떠나버린 것 같다  
+자제 수준이라고도 하고 알고리즘 짤 때에 array = [1,2,3,4] 를 eval('array.join("+")') 이런 식으로 사용했는데  
+eval때문인지는 모르겠지만 성능이 좋지 않았던 경험도 있어서 더 외면하게 되었다
```
eval을 왜 쓰면 안좋은지, 좀더 찾아보면 좋겠네요.

crongro commented on keyword.md:
```diff
@@ -137,6 +162,9 @@ null은 내가 임의로 var a = null 이런식으로 지정해줘야 나오는
 
+__Comment__  
+undefined는 코딩하다 보면 언제든 만나는 녀석이고 이 녀석을 어떻게 관리하고  
+등장 시기를 아느냐에 따라서 실력이 많이 늘어날 것 같다  
```
아무것도 수행하지 못함, 발견하지 못함 등을 표현하고 싶으면 undefiend를 사용하면 돼요.

crongro commented on keyword.md:
```diff
@@ -158,12 +186,15 @@ var module = {
+__Comment__  
+다시 이해하려고 하니 call과 apply랑 똑같은 기능을 하는 것 같은데  
+차이점이 하나 있다 call과 apply는 함수를 바로 실행시켜서 함수의 리턴값을 반환하는데  
+bind는 함수 자체를 반환한다 ()로 실행시켜줘야 하는 셈이다
```
다음미션을 하면서 좀더 자기것으로 이해하도록 해보세요.

crongro commented on keyword.md:
```diff
@@ -184,6 +215,9 @@ speak() //My name is undefined 호출
   
+__Comment__  
+this가 어떤 것인지 대충 이해는 되지만 js에서의 this의 특성까지는 아직 이해가 부족한 것 같다  
+지금까지 이해한 것으로는 인스턴스화 할 때 this가 필수라는 점이다
```
이것도 다음미션을 하면서 이해하기~

crongro commented on keyword.md:
```diff
@@ -216,7 +253,9 @@ function add(v1) {
-
+__Comment__  
+짧은 코딩 인생이지만 fn()()의 패턴은 처음 봤기에 적응이 안됐지만  
+함수형 프로그래밍이기에 가능하다는 것을 깨달았고 함수형 프로그래밍이 굉장히 매력적으로 느껴진다
```
매력을 느끼면 다행이네요~ 
currying 이란 무엇인지 주말에 한번 살펴보세요

crongro reviewed (approved):
정리하느라 수고했어요.
this, call,apply,bind는 중요한데요.
다음미션에서 좀더 이해하면 좋겠네요.

crongro commented on lotto.js:
```diff
@@ -10,84 +10,82 @@ let myLottos = [];
-  FOUR: 50000,
-  FIVE: 1500000,
-  SIX: 2000000000
+  3: 5000,
```
js의 객체는 key값으로 이렇게 숫자를 쓸 수도 있어요. 다른 값도 무엇을 쓸 수 있는지 알아보세요.

crongro commented on lotto.js:
```diff
@@ -10,84 +10,82 @@ let myLottos = [];
 
 // 로또를 넣은 돈만큼 구매
-function buyLottos(money) {
+buyLottos = money => {
```
왜 이녀석은 앞에 const, var, let 등이 없죠?

crongro commented on lotto.js:
```diff
@@ -10,84 +10,82 @@ let myLottos = [];
   let lottoArray = [];
-  for (let i = 0; i < 6; i++) {
+  for (let i = 0; i < num; i++) {
     lottoArray.push(Math.floor(Math.random() * 45) + 1);
```
코드에서 45가 변하는 숫자라고(46이거나 450이거나 등)생각하고 45를 없애보세요.

crongro commented on lotto.js:
```diff
@@ -10,84 +10,82 @@ let myLottos = [];
-    foundSix = 0;
-  let len = myLottos.length;
+setLuckyNumber = array => {
+  const PRIZE_MINIMUM = 3; // 당첨 최소 개수 
```
이 값을 인자로 받는게 더 낫지 않을까요? 최소갯수가 늘 3개라고 생각하지 않는다면요.

crongro commented on lotto.js:
```diff
@@ -10,84 +10,82 @@ let myLottos = [];
+
+  // 로또마다 맞춘 개수를 담아줄 객체, array의 길이만큼 생성
+  let countObj = {}
+  for (let i = PRIZE_MINIMUM; i < array.length + 1; i++) {
```
가급적 변수이름을 array라고 하지 마세요. arr까지는 괜찮고. 좀더 나으려면 lottoList 등이 더 나을 듯.

crongro commented on lotto.js:
```diff
@@ -10,84 +10,82 @@ let myLottos = [];
+  }
+  // 컴퓨터의 번호를 저장할 객체
+  let arrayObj = {};
+  for (let i = 0; i < array.length; i++) {
```
for문말고 배열의 함수형 메서드를 써볼래요?

crongro commented on lotto.js:
```diff
@@ -10,84 +10,82 @@ let myLottos = [];
+  myTotalPrize = Object.keys(countObj).reduce(((prev, curr) => prev + countObj[curr] * PRIZE_MONEY[curr]), 0);
+
+  // 당첨결과를 출력
+  printResult(countObj[3], countObj[4], countObj[5], countObj[6]);
```
이렇게 한번 동작하도록 해보세요.
printResult(countObj, [3,4,5,6]);

crongro reviewed (changes requested):
markdown정리는 괜찮아요. 댓글 참고해서 좀더 학습해보면 좋겠고요. 

lotto는 좀더 수정해보세요.

Commit:
Add 'const' before all functions

Commit:
Modify change max number in function publishnumber

로또의 숫자 최대 범위가 45로 고정적이었는데 max변수를 인자로 받아
가변적으로 사용할 수 있게 변경했습니다.

Commit:
Modify code [리뷰 받은 것 수정]

리뷰 받은 내용대로 수정하고 범용적으로 사용할 수 있도록 조금씩 더 개선했습니다

crongro reviewed (changes requested):
toString 사용이 필요한지?
delete countObj[0]; //아예 안넣기.
reduce를 활용해보세요 ->  lottoArray.push(Math.floor(Math.random() * max) + 1); 

위에 있는 데이터들(mylottos, myTotalPrize, myInvest)를 하나의 객체로 묶기.

console.log(); 지우기 => \n

구현한 코드안에서, 일관된 동작으로 처리되도록 구현하려고 노력. 

countObj = {}는 초기값일 미리 만들필요 없을듯.

주석을 넣는건 좋은데, 약간 줄이면 더 좋겠어요.

Commit:
Delete toString()

Commit:
Modify delete countObj[0]

Commit:
Modify code from review

crongro reviewed (changes requested):
setLuckyNumber 른 나눕시다. 
자세한건 오프라인으로~

Commit:
Delete type in variable name and chunk setLuckyNumber

crongro commented on lotto.js:
```diff
@@ -39,31 +39,36 @@ const publishNumber = (num = 6, max = 45) => {
+const savaCorrectedNumber = (computerNumber) => {
+  let countedNumber = {};
+  let computerNumberFlag = {};
+  computerNumber.reduce((prev, curr) => computerNumberFlag[curr] = true, {});
```
reduce는 새로운 어떤 value를 반환하는 함수인데, 반환한 결과를 밖에 변수에 저장하고 있어요.
var result = [].reduce();
이렇게 돼야 할 듯.

crongro commented on lotto.js:
```diff
@@ -39,31 +39,36 @@ const publishNumber = (num = 6, max = 45) => {
+// 맞춘 개수대로 선언한 객체에 저장
+const savaCorrectedNumber = (computerNumber) => {
+  let countedNumber = {};
+  let computerNumberFlag = {};
```
computerNumberFlag 이것을 꼭 만들어야 할까요? 중간에 불필요한 객체를 하나 만드는 거 같아요.
이걸 안쓰고 해결해보세요.

crongro commented on lotto.js:
```diff
@@ -6,7 +6,7 @@
 
 
-let lottoObj = {
+let lotto = {
```
lotto 를 전역공간에 두고 있는데요.
전역공간을 없애고, 함수끼리 서로 인자로 전달해주면서 처리하도록 한번 해볼래요?

crongro reviewed (changes requested):
함수의 규모가 작아지면서, 복잡함이 조금 줄어들었네요.
불필요해보이는 부분을 조금더 줄일 수 있을 거 같아요.

Commit:
Modify saveCorrectNumber function simply

1. computerNumberFlag객체를 만들어 객체의 key값이 있을 때 내 숫자의 맞춘개수를
카운트 하던 방식을 사용했어서 computerNumberFlag라는 객체를 또 하나 만들었었는데
그냥 다이렉트로 배열의 indexOf를 사용해서 하는 방식으로 변화시켰다
그래서 checkNumber의 인자로 객체가 아닌 computerNumber를 그대로 넘겨주고 그 안에서 indexOf를 동작시킨다

2. 이로 인해서 리뷰남겨주신 reduce를 이요하여 반환값이 아닌 외부 값만 변화시키는 변수 자체가 없어졌다

Commit:
Delete global object lotto

crongro reviewed (approved):
네 로또는 여기까지 하는걸로 해요.
수고했어요.

추가로, 다음부터는 
함수이름은 동사+명사로, 
변수는 명사로(줄여쓰지말기)
잘 만들도록 하세요.

LOG91 commented:
네 감사합니다 크롱!
